main.go

```diff
@@ -119,47 +119,110 @@ func newMemo() *memo {
 }
 
 func (m *memo) build(n *node) {
-	switch n.typ {
-	case join:
+	switch n.op.typ() {
+	case joinNode:
 		m.build(n.left)
 		m.build(n.right)
 		m.add(n)
 
-	case scan:
+	case scanNode:
 		m.add(n)
 	}
 }
 
+type xform interface {
+	check(n *node) bool
+	apply(n *node) []*node
+}
+
 // A ⋈ B => B ⋈ A
-func (m *memo) commute(n *node) *node {
-	if n.typ != join {
+type joinCommuteXform struct{}
+
+func (jc *joinCommuteXform) check(n *node) bool {
+	if n.op.typ() != joinNode {
+		return false
+	}
+	return true
```

Comment:
Nit: I would have defined this as `return n.op.typ() == joinNode`. Shorter code, but also a bit easier to read the positive condition.